src/script/src/python/ffi_types/vector.rs

```diff
@@ -0,0 +1,524 @@
+            Ok(v.into_pyobject(vm))
+        }
+    }
+
+    /// Unsupported
+    /// TODO(discord9): make it work
+    #[allow(unused)]
+    fn setitem_by_index(
+        zelf: PyRef<Self>,
+        i: isize,
+        value: PyObjectRef,
+        vm: &VirtualMachine,
+    ) -> PyResult<()> {
+        Err(vm.new_not_implemented_error("setitem_by_index unimplemented".to_string()))
+    }
+
+    /// rich compare, return a boolean array, accept type are vec and vec and vec and number
+    pub(crate) fn richcompare(
+        &self,
+        other: PyObjectRef,
+        op: PyComparisonOp,
+        vm: &VirtualMachine,
+    ) -> PyResult<PyVector> {
+        if rspy_is_pyobj_scalar(&other, vm) {
+            let scalar_op = get_arrow_scalar_op(op);
+            self.rspy_scalar_arith_op(other, None, scalar_op, vm)
+        } else {
+            let arr_op = get_arrow_op(op);
+            self.rspy_vector_arith_op(other, None, wrap_result(arr_op), vm)
+        }
+    }
+
+    pub(crate) fn len(&self) -> usize {
+        self.as_vector_ref().len()
+    }
+}
+
+/// get corresponding arrow op function according to given PyComaprsionOp
+fn get_arrow_op(op: PyComparisonOp) -> impl Fn(&dyn Array, &dyn Array) -> ArrowResult<ArrayRef> {
+    let op_bool_arr = match op {
+        PyComparisonOp::Eq => comparison::eq_dyn,
+        PyComparisonOp::Ne => comparison::neq_dyn,
+        PyComparisonOp::Gt => comparison::gt_dyn,
+        PyComparisonOp::Lt => comparison::lt_dyn,
+        PyComparisonOp::Ge => comparison::gt_eq_dyn,
+        PyComparisonOp::Le => comparison::lt_eq_dyn,
+    };
+
+    move |a: &dyn Array, b: &dyn Array| -> ArrowResult<ArrayRef> {
+        let array = op_bool_arr(a, b)?;
+        Ok(Arc::new(array))
+    }
+}
+
+/// get corresponding arrow scalar op function according to given PyComaprsionOp
+fn get_arrow_scalar_op(
+    op: PyComparisonOp,
+) -> impl Fn(&dyn Array, &dyn Array) -> Result<ArrayRef, String> {
+    let op_bool_arr = match op {
+        PyComparisonOp::Eq => comparison::eq_dyn,
+        PyComparisonOp::Ne => comparison::neq_dyn,
+        PyComparisonOp::Gt => comparison::gt_dyn,
+        PyComparisonOp::Lt => comparison::lt_dyn,
+        PyComparisonOp::Ge => comparison::gt_eq_dyn,
+        PyComparisonOp::Le => comparison::lt_eq_dyn,
+    };
+
+    move |a: &dyn Array, b: &dyn Array| -> Result<ArrayRef, String> {
+        let array = op_bool_arr(a, b).map_err(|e| format!("scalar op error: {e}"))?;
+        Ok(Arc::new(array))
+    }
+}
+
+/// if this pyobj can be cast to a scalar value(i.e Null/Int/Float/Bool)
+#[inline]
+pub(crate) fn rspy_is_pyobj_scalar(obj: &PyObjectRef, vm: &VirtualMachine) -> bool {
+    is_instance::<PyNone>(obj, vm)
+        || is_instance::<PyInt>(obj, vm)
+        || is_instance::<PyFloat>(obj, vm)
+        || is_instance::<PyBool>(obj, vm)
+}
+
+/// convert a DataType `Value` into a `PyObjectRef`
+pub fn val_to_pyobj(val: value::Value, vm: &VirtualMachine) -> PyObjectRef {
+    match val {
```

Comment:
FYI: #1061 will add `Dictionary` type, which may break this match

Reply:
> FYI: #1061 will add `Dictionary` type, which may break this match

Should Value be added a  `#[non_exhaustive]` attr to always cover wildcard arm?